internal/report/snapshot_reporter.go

```diff
@@ -0,0 +1,214 @@
+// Copyright  observIQ, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package report contains reporters for collecting specific information about the collector-
+package report
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+
+	"github.com/observiq/observiq-otel-collector/internal/report/snapshot"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+)
+
+// snapShotKind is the kind for the snapshot reporter
+var snapShotKind = "snapshot"
+
+// snapshotConfig specifies what snapshots to collect
+type snapshotConfig struct {
+	// Endpoint is where to send the snapshots
+	Endpoint *endpointConfig `yaml:"endpoint"`
+
+	// Processor is the full ComponentID of the snapshot processor
+	Processor string `yaml:"processor"`
+
+	// PipelineType will be "logs", "metrics", or "traces"
+	PipelineType string `yaml:"pipeline_type"`
+}
+
+// endpointConfig is the configuration of a specific endpoint and full headers to include
+type endpointConfig struct {
+	URL     string      `yaml:"url"`
+	Headers http.Header `yaml:"headers"`
+}
+
+var _ Reporter = (*SnapshotReporter)(nil)
+var _ snapshot.Snapshotter = (*SnapshotReporter)(nil)
+
+// SnapshotReporter tracks and reports snapshots
+type SnapshotReporter struct {
+	client Client
+
+	// minPayloadSize is the minimum number of items to be in a payload
+	minPayloadSize int
+
+	// Buffers
+	logBuffers    map[string]*snapshot.LogBuffer
+	metricBuffers map[string]*snapshot.MetricBuffer
+	traceBuffers  map[string]*snapshot.TraceBuffer
+}
+
+// NewSnapshotReporter creates a new SnapshotReporter with the associated client
+func NewSnapshotReporter(client Client) *SnapshotReporter {
+	return &SnapshotReporter{
+		client:         client,
+		minPayloadSize: 100,
+		logBuffers:     make(map[string]*snapshot.LogBuffer),
+		metricBuffers:  make(map[string]*snapshot.MetricBuffer),
+		traceBuffers:   make(map[string]*snapshot.TraceBuffer),
+	}
+}
+
+// Kind returns kind of the reporter
+func (s *SnapshotReporter) Kind() string {
+	return snapShotKind
+}
+
+// Report applies the new configuration and reports snapshots specified in it
+func (s *SnapshotReporter) Report(cfg any) error {
+	ssCfg, ok := cfg.(*snapshotConfig)
+	if !ok {
+		return errors.New("invalid config type")
+	}
```

Comment:
I think this check should be hoisted to the highest level. Actually, this is originally unmarshalled into a snapshotConfig earlier on in the Manager, so might as well just keep the cfg type as `snapshotConfig` from that point onwards instead of turning it into `any`.

Reply:
In order to fulfill the reporter interface I need to pass an `any` so that I can pass in any config type and the reporter knows which type of config it needs. We should never pass in the wrong type based on the logic flow but it's an extra protection.